### DISCUSSION TITLE
How to use MPI to train distributed?

### DISCUSSION THREAD

**Discussion by FDInSky** [open-mmlab/mmdetection#8978] [Q&A] [answered]
https://github.com/open-mmlab/mmcv/blob/8cac7c25ee5bc199d6e4059297ef2fa92d9c069c/mmcv/runner/dist_utils.py#L35

How to use mpi to train distributed?

**Comment by ZwwWayne** [comment]
You can set the launcher equal to mpi, and use mpirun to launch the code.

**Comment by hyz-xmaster** [comment]
@ZwwWayne , hi, I was also wondering how to use MPI for multi-machine training. Could you give an example here?

**Comment by yingfhu** [comment, marked as answer]
> @ZwwWayne , hi, I was also wondering how to use MPI for multi-machine training. Could you give an example here?

An example for use MPI run for distributed training with 2 GPUS on 2 nodes (1 GPU per node).
```shell
mpirun \
--allow-run-as-root \
--npernode 1 --np 2 \
python tools/train.py ${CONFIG_FILE} --launcher mpi
```

Note: Should at least set `MASTER_ADDR` environment variable which is necessary for pytorch. Refers to
https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/dist_utils.py#L66

**Comment by hyz-xmaster** [comment]
@yingfhu Great, thanks for the help.
